### PR TITLE
AnyCPU (run as either 64-bit or 32-bit as needed)

### LINKED ISSUE #12587: Switch to 64-bit executables?
Modern games and operating systems tend to be 64-bit only these days. How about we do that switch as well? My hope is that it will address issues like https://github.com/OpenRA/OpenRA/issues/12494 as we gain more access to memory and maybe it will improve overall performance. Is that too naive? My technical knowledge in that regard is limited.

Porting should be easy. For the managed code just flip some switches and for the system calls we only need to update the bundled native Windows/Mac dependencies to their 64 bit counterparts and for Linux just update package architecture metadata.

I haven't checked how many players are still running 32-bit operating systems. Is that possible from our player survey? Hopefully no one will be excluded. My assumption or rather wishful thinking would be that most of them could upgrade to a 64-bit OS for their own benefit and 32-bit only era hardware will also make playing OpenRA especially the upcoming TS/RA2 mods not much fun anyway.

### PR DESCRIPTION
**Pull request by RoosterDragon**
Fixes #12587. Changes our build configuration from x86 (32-bit) to AnyCPU. This allows OpenRA to run on either 32-bit or 64-bit systems without needing two different builds.

- I have made other changes required so this works on Windows. Please let me know if other platforms also require changes of any sort, or if I missed something that needs to be tweaked in the build system so this builds correctly on other platforms.
- If testing on Windows, you may need to remove copies of the following native DLLs from the build output: `freetype6.dll`, `lua51.dll`, `SDL2.dll`, `soft_oal.dll`. This branch no longer copies these, but if they are already present in your output before you switch, they may create an issue when you try and run it.

Testing in a local game with a 32-bit and 64-bit client I did not see any desyncs. We'll want to do more testing for sure, but it doesn't immediately explode at least. :)

### CODE REVIEW

File: OpenRA.Game/Support/WindowsPreparedNativeDll.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+			// We could copy the correct DLL locally, but this fails if OpenRA is installed in a directory protected by
+			// UAC. Instead we are forced to create a temporary directory and copy the DLL there. We then need to add
+			// the temporary directory to the PATH in order to convince the loader to look there when required.
+			// This is dumb, but it works.
+			tempDirectory = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString());
+			Directory.CreateDirectory(tempDirectory);
+			var filename = Path.GetFileNameWithoutExtension(file);
+			var ext = Path.GetExtension(file);
+			var sourceFile = filename + (Environment.Is64BitProcess ? "-x64" : "-x86") + ext;
+			File.Copy(sourceFile, Path.Combine(tempDirectory, file), true);
```
Comment (pchote) [inline OpenRA.Game/Support/WindowsPreparedNativeDll.cs]: IMO this is really quite nasty.  It looks like it is feasible to create a short stub to dynamically load native libraries (see e.g. http://andyrushton.co.uk/csharp-dynamic-loading-of-a-c-dll-at-run-time/) - if we don't mind maintaining our own tweaked forks of sharpfont, SDL2-CS etc then we could handle this there and have the engine Just Work™

Comment (pchote) [reply OpenRA.Game/Support/WindowsPreparedNativeDll.cs]: This is a pretty big choice either ways, so we should get more opinions before we decide on a direction forwards.

Comment (pchote) [reply OpenRA.Game/Support/WindowsPreparedNativeDll.cs]: Another option could be to maintain our own minimal bindings in-project, like we do [for OpenGL](https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Platforms.Default/OpenGL.cs)

**Comment by RoosterDragon** [comment]
https://github.com/OpenRA/OpenRA/pull/14675#discussion_r160031319 (so it doesn't get lost as changes are made)

File: thirdparty/fetch-thirdparty-deps-windows.sh [hunk 2]
```diff
@@ -8,30 +8,36 @@ download_dir="${0%/*}/download/windows"
 mkdir -p "${download_dir}"
 cd "${download_dir}"
 
-if [ ! -f SDL2.dll ]; then
+if [ ! -f SDL2-x86.dll -o ! -f SDL2-x64.dll ]; then
 	echo "Fetching SDL2 from libsdl.org"
 	curl -LOs https://www.libsdl.org/release/SDL2-2.0.5-win32-x86.zip
-	unzip SDL2-2.0.5-win32-x86.zip SDL2.dll
+	unzip SDL2-2.0.5-win32-x86.zip SDL2-x86.dll
```
Comment (RoosterDragon) [inline thirdparty/fetch-thirdparty-deps-windows.sh]: Looks like I need to unzip and then rename - how do I spell that out?

Comment (pchote) [reply thirdparty/fetch-thirdparty-deps-windows.sh]: Add a separate line `mv SDL2.dll SDL2-x86.dll`

**Comment by LipkeGu** [comment]
We could also use native opengl libaries (Windows, linux and OSX) ships them).

**Comment by LipkeGu** [comment]
btw this PR includes all our  fixheader changes at compile time (/LAA)
this flag is set by the compiler.

File: thirdparty/fetch-thirdparty-deps-windows.sh [hunk 3]
```diff
@@ -8,30 +8,38 @@ download_dir="${0%/*}/download/windows"
...
 
-if [ ! -f SDL2.dll ]; then
+if [ ! -f SDL2-x86.dll -o ! -f SDL2-x64.dll ]; then
 	echo "Fetching SDL2 from libsdl.org"
 	curl -LOs https://www.libsdl.org/release/SDL2-2.0.5-win32-x86.zip
 	unzip SDL2-2.0.5-win32-x86.zip SDL2.dll
+	mv SDL2.dll SDL2-x86.dll
 	rm SDL2-2.0.5-win32-x86.zip
+	curl -LOs https://www.libsdl.org/release/SDL2-2.0.5-win32-x64.zip
+	unzip SDL2-2.0.5-win32-x86.zip SDL2.dll
```
Comment (Mailaender) [inline thirdparty/fetch-thirdparty-deps-windows.sh]: Small oversight. This should be:

```patch
-unzip SDL2-2.0.5-win32-x86.zip SDL2.dll
+unzip SDL2-2.0.5-win32-x64.zip SDL2.dll
```

**Comment by pchote** [comment]
We discussed this in IRC when it was first opened, and decided that it would be better to shim the dll loading instead of copying and executing the game code from a temporary directory.  If that isn't feasible then going 64-bit only remains as a lower maintenance option.

**Comment by RoosterDragon** [comment]
I didn't think of it at the time - but I think one problem with shimming the DLLs is that we'd still have to have differently named DLLs, and that might require messing with the dllmap so things still worked on non-Windows. That's a mess I don't think I want to get into.

A few options:

Method | Advantages | Disadvantages
--- | --- | ---
AnyCPU with this PR | 32 & 64 bit support in one binary | Hacky loading from a temp dir on Windows
AnyCPU with shimmed DLLs | 32 & 64 bit support in one binary | Effort
64 bit only | 64 bit support | Drops 32 bit support on Windows
32 and 64 bit builds | 32 & 64 bit support in two binaries | Two Windows downloads to choose from

I like this PR for the single binary benefit - but if the hacky bit is too much, I think the next best option is 32 & 64 bit Windows downloads. That way we don't have to drop support for 32-bit Windows. The only negative is the support if people get confused over the two download options.

File: thirdparty/fetch-thirdparty-deps-windows.sh (hunk 3 above)
Comment (RoosterDragon) [reply thirdparty/fetch-thirdparty-deps-windows.sh]: Thanks for spotting this - I was going blind trying to work out what was wrong.

**Comment by pchote** [comment]
Could we do the 32/64bit check in the installer, and install the appropriate versions of the dll once.  This would give the benefits of separate builds without the confusion of two downloads.  The `make dependencies` script could even more easily do the same for development code.

**Comment by RoosterDragon** [comment]
We would need the installer to also reach into the exe and fiddle the header to indicate if it wants 32 or 64 bit or not. Similarly for development you'd need to switch between a 32/64 bit build (or if it's an AnyCPU build, toggle the "Prefer 32-bit" flag). But it's still a reasonable approach.

It's definitely beyond my current motivation to muck with the installer and try all that out - I'll close this for now. If anybody is feeling motivated enough to muck with the installer then please feel free to have a go. :)

**Comment by pchote** [comment]
Superseded by #16580 (🎉), so removing from the milestone

**Pull request closed without merging**